### PR TITLE
[TwigBundle] Enhance the twig not found exception

### PR DESCRIPTION
**Pull request by behnoushnorouzi**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | no 
| Tests pass?   | yes 
| Fixed tickets | #27468
| License       | MIT
| Doc PR        | -

Improve error message to make it clear to developers what mistake they have made.

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function findTemplate($template, $throw = true)
+    {
+        try {
+            return parent::findTemplate($template, $throw);
+        } catch (LoaderError $e) {
+            if ($e instanceof \Twig_Error_Loader && preg_match('/^([^:]*?)(?:Bundle)?:([^:]*+):(.+\.[^\.]+\.[^\.]+)$/', $template, $m) && 0 !== strpos($template, '@')) {
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php]: the `instanceof` will always be true. `Twig\Error\LoaderError` is a class alias of `Twig_Error_Loader`.

Comment (stof) [inline src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php]: You should switch the regex match in the `strpos`. Checking the first char is faster.

Btw, you could optimize it even more:

```php
if ('' !== $template && '@' !== $template[0] && preg_match(...)) {
```

Using strpos to check only the first char is not the fastest way.

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Bundle/TwigBundle/Tests/Loader/NativeFilesystemLoaderTest.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+namespace Symfony\Bundle\TwigBundle\Tests\Loader;
+
+use Symfony\Bundle\TwigBundle\Loader\NativeFilesystemLoader;
+use Symfony\Bundle\TwigBundle\Tests\TestCase;
+
+class NativeFilesystemLoaderTest extends TestCase
+{
+    const DS = DIRECTORY_SEPARATOR;
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/Tests/Loader/NativeFilesystemLoaderTest.php]: I would remove this const and use the PHP one instead.

File: src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function findTemplate($template, $throw = true)
+    {
+        try {
+            return parent::findTemplate($template, $throw);
+        } catch (LoaderError $e) {
+            if ('' === $template || '@' === $template[0] || !preg_match('/^([^:]*?)(?:Bundle)?:([^:]*+):(.+\.[^\.]+\.[^\.]+)$/', $template, $m)) {
```
Comment (fabpot) [inline src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php]: Can we use named patterns to make the code using `$m` more easily understandable?

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by behnoushnorouzi** [comment]
Thank you for the reviews it's done.

Status: needs review

**Review by sstok — APPROVED** [review]

File: src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\TwigBundle\Loader;
+
+use Twig\Error\LoaderError;
+use Twig\Loader\FilesystemLoader;
+
+class NativeFilesystemLoader extends FilesystemLoader
```
Comment (xabbuh) [inline src/Symfony/Bundle/TwigBundle/Loader/NativeFilesystemLoader.php]: you may want to add yourself as the author of this class :)

**Comment by Lewiscowles1986** [comment]
Should this emit a notice or warning if used in production? 

It seems like a development pattern, and can be switched out for the extended class. Once the suggestion is implemented it seems like switching to the standard class might be more helpful.

**Review by xabbuh — APPROVED** [review]

**Review by 20uf — APPROVED** [review]

**Comment by stof** [comment]
We don't even need to do it in a compiler pass. The debug parameter can be accessed in the DI extension (the `kernel.*` params can be accessed there safely, as they are set early)

**Comment by nicolas-grekas** [comment]
Thank you @behnoushnorouzi.

**Pull request merged**